Content.Server/Magic/MagicSystem.cs

```diff
@@ -0,0 +1,321 @@
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        foreach (var forcewall in EntityQuery<ForceWallSpellComponent>())
+        {
+            forcewall.Lifetime -= frameTime;
+
+            if (forcewall.Lifetime <= 0)
+                EntityManager.QueueDeleteEntity(forcewall.Owner);
+        }
+
+        foreach (var spawnedEntities in EntityQuery<SpawnSpellComponent>())
+        {
+            spawnedEntities.Lifetime -= frameTime;
+
+            if (spawnedEntities.Lifetime <= 0)
+                EntityManager.QueueDeleteEntity(spawnedEntities.Owner);
+        }
+    }
+
+    private void OnUse(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        AttemptLearn(uid, component, args);
+
+        args.Handled = true;
+    }
+
+    private void AttemptLearn(EntityUid uid, SpellbookComponent component, UseInHandEvent args)
+    {
+        component.CancelToken = new();
+        var doAfterEventArgs = new DoAfterEventArgs(args.User, component.LearnTime, component.CancelToken.Token, uid)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            BreakOnDamage = true,
+            BreakOnStun = true,
+            NeedHand = true, //What, are you going to read with your eyes only??
+            TargetFinishedEvent = new LearnDoAfterComplete(args.User),
+            TargetCancelledEvent = new LearnDoAfterCancel(),
+        };
+
+        _doAfter.DoAfter(doAfterEventArgs);
+    }
+
+    private void TeachSpell(EntityUid uid, SpellbookComponent component, LearnDoAfterComplete ev)
+    {
+        component.CancelToken = null;
+
+        if (ev.User == null)
+            return;
+
+        _actionsSystem.AddActions(ev.User, component.Spells, uid);
+    }
+
+    #region Spells
+
+    private void OnRuneMagic(RuneMagicEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+        Spawn(args.RunePrototype, transform.Coordinates.SnapToGrid());
+
+        args.Handled = true;
+    }
+
+    /// <summary>
+    /// Teleports the user to the clicked location
+    /// </summary>
+    /// <param name="args"></param>
+    private void OnTeleportSpell(TeleportSpellEvent args)
+    {
+        if (args.Handled)
+            return;
+
+        var transform = Transform(args.Performer);
+
+        if (_mapManager.TryFindGridAt(args.Target, out var grid))
+        {
+            var gridPosition = grid.WorldToLocal(args.Target.Position);
+
+            transform.Coordinates = new EntityCoordinates(grid.GridEntityId, gridPosition);
+            SoundSystem.Play(Filter.Pvs(args.Target), args.BlinkSound.GetSound());
+        }
+        else
+        {
+            var mapEntity = _mapManager.GetMapEntityIdOrThrow(args.Target.MapId);
+            transform.WorldPosition = args.Target.Position;
+            transform.AttachParent(mapEntity);
```

Comment:
```suggestion
            transform.AttachParent(mapEntity);
            transform.WorldPosition = args.Target.Position;
```
When setting position, make sure you do it after AttachParent because AttachParent usually moves the entity to 0,0 of the parent.